Sources/GRPC/GRPCClientStateMachine.swift

```diff
@@ -0,0 +1,588 @@
+    return self.state.recieveEndOfResponseStream(trailers)
+  }
+}
+
+extension GRPCClientStateMachine.State {
+  /// See `GRPCClientStateMachine.sendRequestHeaders(host:path:options:requestID)`.
+  mutating func sendRequestHeaders(
+    host: String,
+    path: String,
+    options: CallOptions,
+    requestID: String
+  ) -> Result<HTTPRequestHead, InvalidStateError> {
+    let result: Result<HTTPRequestHead, InvalidStateError>
+
+    switch self {
+    case let .clientIdleServerIdle(pendingWriteState, responseArity):
+      let head = self.makeRequestHeaders(host: host, path: path, options: options, requestID: requestID)
+      result = .success(head)
+      self = .clientStreamingServerIdle(pendingWriteState.makeWriteState(), responseArity)
+
+    case .clientStreamingServerIdle,
+         .clientClosedServerIdle,
+         .clientClosedServerStreaming,
+         .clientStreamingServerStreaming,
+         .clientClosedServerClosed:
+      result = .failure(.invalidState())
+    }
+
+    return result
+  }
+
+  /// See `GRPCClientStateMachine.sendRequest(_:allocator:)`.
+  mutating func sendRequest(
+    _ message: Request,
+    allocator: ByteBufferAllocator
+  ) -> Result<ByteBuffer, MessageWriteError> {
+    let result: Result<ByteBuffer, MessageWriteError>
+
+    switch self {
+    case .clientStreamingServerIdle(var writeState, let arity):
+      result = writeState.write(message, allocator: allocator)
+      self = .clientStreamingServerIdle(writeState, arity)
+
+    case .clientStreamingServerStreaming(var writeState, let readState):
+      result = writeState.write(message, allocator: allocator)
+      self = .clientStreamingServerStreaming(writeState, readState)
+
+    case .clientClosedServerIdle,
+         .clientClosedServerStreaming,
+         .clientClosedServerClosed:
+      result = .failure(.cardinalityViolation)
+
+    case .clientIdleServerIdle:
+      result = .failure(.invalidState)
+    }
+
+    return result
+  }
+
+  /// See `GRPCClientStateMachine.sendEndOfRequestStream()`.
+  mutating func sendEndOfRequestStream() -> Result<Void, SendEndOfRequestStreamError> {
+    let result: Result<Void, SendEndOfRequestStreamError>
+
+    switch self {
+    case .clientStreamingServerIdle(_, let responseArity):
+      result = .success(())
+      self = .clientClosedServerIdle(responseArity)
+
+    case .clientStreamingServerStreaming(_, let readState):
+      result = .success(())
+      self = .clientClosedServerStreaming(readState)
+
+    case .clientClosedServerIdle,
+         .clientClosedServerStreaming,
+         .clientClosedServerClosed:
+      result = .failure(.endOfStreamCardinalityViolation)
+
+    case .clientIdleServerIdle:
+      result = .failure(.invalidState)
+    }
+
+    return result
+  }
+
+  /// See `GRPCClientStateMachine.recieveResponseHeaders(_:)`.
+  mutating func recieveResponseHeaders(
+    _ responseHead: HTTPResponseHead,
+    logger: Logger
+  ) -> Result<HTTPHeaders, RecieveResponseHeadError> {
```

Comment:
Make `Head/Headers` consistent in `receiveResponseHeaders` and `ReceiveResponseHeadError`?

Reply:
Not done yet, I think.